app/controllers/general.php

```diff
@@ -40,9 +43,136 @@
 Config::setParam('cookieDomain', 'localhost');
 Config::setParam('cookieSamesite', Response::COOKIE_SAMESITE_NONE);
 
+function router(App $utopia, Database $dbForConsole, SwooleRequest $swooleRequest, Request $request, Response $response)
+{
+    $utopia->getRoute()->label('error', __DIR__ . '/../views/general/error.phtml');
+
+    $host = $request->getHostname() ?? '';
+
+    $route = Authorization::skip(
+        fn () => $dbForConsole->find('rules', [
+            Query::equal('domain', [$host]),
+            Query::limit(1)
+        ])
+    )[0] ?? null;
+
+    if ($route === null) {
+        $mainDomain = App::getEnv('_APP_DOMAIN', '');
+
+        if ($mainDomain === 'localhost') {
+            throw new AppwriteException(AppwriteException::GENERAL_SERVER_ERROR, 'Please configure domain environment variables before using Appwrite outside of localhost.');
+        } else {
+            throw new AppwriteException(AppwriteException::ROUTER_HOST_NOT_FOUND);
+        }
+    }
+
+    $projectId = $route->getAttribute('projectId');
+    $project = Authorization::skip(
+        fn () => $dbForConsole->getDocument('projects', $projectId)
+    );
```

Comment:
can we not get the project using the `project` resource in utopia? 
https://github.com/appwrite/appwrite/blob/master/app/init.php#L901-L915

Reply:
Nope. Project is from header or URL param project ID.
With custom and preview domains, you don't have project ID from request, instead, you read it from rule document.